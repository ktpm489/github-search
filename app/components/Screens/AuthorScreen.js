import React from 'react';
import { View, Text, StyleSheet, ActivityIndicator } from 'react-native';
import { graphql, compose } from 'react-apollo';
import { WebBrowser } from 'expo';

import RepositoryOwner from '../Repository/RepositoryOwner';
import AuthorOverview from '../Author/AuthorOverview';
import getSearch from '../../graphql/getSearch';
import getAuthor from '../../graphql/getAuthor';

const openProfile = async (data) => {
  await WebBrowser.openBrowserAsync(data.user.url);
};

export const AuthorScreen = ({ data }) => {
  if (data.loading) {
    return (
      <View style={styles.container}>
        <ActivityIndicator size="large" color="#000" />
      </View>
    );
  }

  return (
    <View style={styles.container}>
      <RepositoryOwner
        url={data.user.url}
        avatarUrl={data.user.avatarUrl}
        name={data.user.name}
        login={data.user.login}
        containerStyle={{ marginBottom: 10 }}
        infoContainerStyle={styles.infoContainer}
        onAvatarPress={() => openProfile(data)}
      />
      {!!data.user.bio && <Text style={styles.bio}>{data.user.bio}</Text>}
      <AuthorOverview
        company={data.user.company}
        location={data.user.location}
        url={data.user.url}
        followerCount={data.user.followers.totalCount}
        websiteUrl={data.user.websiteUrl}
        repoCount={data.user.repositories.totalCount}
        organizations={data.user.organizations.edges}
      />
    </View>
  );
};

export default compose(
  graphql(getSearch, {
    props: ({
      data: {
        search: { input },
      },
    }) => ({
      input,
    }),
  }),
  graphql(getAuthor, {
    options: ({ input }) => ({
      variables: {
        login: input,
      },
    }),
  })
)(AuthorScreen);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    padding: 20,
    justifyContent: 'space-around',
  },
  infoContainer: {
    alignItems: 'center',
    marginBottom: 20,
  },
  name: {
    color: '#000',
    fontSize: 22,
    fontWeight: '700',
  },
  login: {
    color: '#6C7680',
    fontSize: 18,
  },
  bio: {
    color: '#000',
    fontSize: 18,
    marginBottom: 20,
  },
});
