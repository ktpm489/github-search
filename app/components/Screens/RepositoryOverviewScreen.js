import React from 'react';
import { graphql, compose } from 'react-apollo';
import { StyleSheet, ActivityIndicator } from 'react-native';
import Error from '../Primitives/Error';
import ScreenView from '../Primitives/ScreenView';
import RepositoryOwner from '../Repository/RepositoryOwner';
import RepositoryOverviewList from '../Repository/RepositoryOverviewList';
import getRepositoryOverview from '../../graphql/getRepositoryOverview';
import getSearch from '../../graphql/getSearch';

class RepositoryOverviewScreen extends React.Component {
  onFetchMore = () => {
    const {
      data: {
        fetchMore,
        repositoryOwner: {
          repositories: { edges },
        },
      },
    } = this.props;
    const cursor = edges[edges.length - 1].cursor;

    fetchMore({
      variables: {
        cursor,
      },
      updateQuery: (previousResult, { fetchMoreResult }) => {
        const prevRepos = previousResult.repositoryOwner.repositories.edges;
        const newRepos = fetchMoreResult.repositoryOwner.repositories.edges;
        const edges = [...prevRepos, ...newRepos];

        return {
          ...previousResult,
          repositoryOwner: {
            ...previousResult.repositoryOwner,
            repositories: {
              ...previousResult.repositoryOwner.repositories,
              edges,
            },
          },
        };
      },
    });
  };

  render() {
    const {
      data: { loading, error, repositoryOwner, user },
      navigation,
    } = this.props;

    if (loading) {
      return (
        <ScreenView>
          <ActivityIndicator size="large" color="#222" />
        </ScreenView>
      );
    }

    if (!repositoryOwner || error) {
      return <Error navigation={navigation} />;
    }

    return (
      <ScreenView>
        <RepositoryOwner
          url={repositoryOwner.url}
          avatarUrl={repositoryOwner.avatarUrl}
          name={user.name}
          login={user.login}
          height={75}
          width={75}
          flexDirection="row"
          infoContainerStyle={styles.infoContainer}
          onAvatarPress={() => navigation.navigate('AuthorScreen')}
        />
        <RepositoryOverviewList
          repos={repositoryOwner.repositories.edges}
          navigation={navigation}
          onEndReached={this.onFetchMore}
        />
      </ScreenView>
    );
  }
}

export default compose(
  graphql(getSearch, {
    props: ({
      data: {
        search: { input },
      },
    }) => ({
      input,
    }),
  }),
  graphql(getRepositoryOverview, {
    options: ({ input }) => ({
      variables: {
        login: input,
        cursor: null,
      },
      errorPolicy: 'all',
    }),
  })
)(RepositoryOverviewScreen);

const styles = StyleSheet.create({
  infoContainer: {
    flex: 1,
    marginLeft: 10,
    justifyContent: 'center',
  },
});
