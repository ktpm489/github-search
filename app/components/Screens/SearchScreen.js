import React from 'react';
import {
  KeyboardAvoidingView,
  View,
  Image,
  Text,
  StyleSheet,
  Platform,
  Keyboard,
} from 'react-native';
import TextInput from '../Primitives/TextInput';
import { Button } from 'react-native-elements';
import { graphql, compose } from 'react-apollo';
import getSearch from '../../graphql/getSearch';
import trackSearch from '../../graphql/trackSearch';

class SearchScreen extends React.Component {
  onSubmitHandler = () => {
    const { navigation, input } = this.props;

    // dismiss the keyboard and navigate to the next screen
    Keyboard.dismiss();
    navigation.navigate('RepositoryOverviewScreen', { username: input });
  };

  onChangeText = (text) => {
    const { handleChange } = this.props;
    handleChange(text.trim().toLowerCase());
  };

  render() {
    const { input } = this.props;

    return (
      <KeyboardAvoidingView style={styles.container} behavior="padding">
        <View style={styles.titleContainer}>
          <Image
            style={{ width: 100, height: 40 }}
            source={require('../../../assets/GitHub_Logo.png')}
          />
          <Text style={styles.title}>Search</Text>
        </View>
        <Image
          style={{ width: 100, height: 100, marginBottom: 20 }}
          source={require('../../../assets/GitHub-Mark-120px-plus.png')}
        />
        <TextInput
          onChangeText={this.onChangeText}
          value={input}
          placeholder="Search for a GitHub user"
          underlineColorAndroid="transparent"
        />
        <Button
          small
          icon={{ name: 'search' }}
          title="Search"
          onPress={this.onSubmitHandler}
          backgroundColor="#222"
          buttonStyle={styles.button}
        />
      </KeyboardAvoidingView>
    );
  }
}

export default compose(
  graphql(getSearch, {
    props: ({
      data: {
        search: { input },
      },
    }) => ({
      input,
    }),
  }),
  graphql(trackSearch, {
    props: ({ mutate }) => ({
      handleChange: (input) => mutate({ variables: { input } }),
    }),
  })
)(SearchScreen);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 20,
  },
  titleContainer: {
    flexDirection: 'row',
  },
  title: {
    fontSize: 28,
    fontFamily: Platform.OS === 'ios' ? 'Avenir Next' : 'Roboto',
  },
  button: {
    marginTop: 10,
    paddingLeft: 20,
    paddingRight: 20,
  },
});
