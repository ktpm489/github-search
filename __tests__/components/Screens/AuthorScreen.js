import React from 'react';
import { ActivityIndicator } from 'react-native';
import { shallow } from 'enzyme';

import { AuthorScreen } from '../../../app/components/Screens/AuthorScreen';

describe('<AuthorScreen />', () => {
  const props = {
    data: {
      user: {
        url: 'https://github.com/parkerziegler',
        avatarUrl: 'https://github.com/parkerziegler/avatar.png',
        name: 'Parker Ziegler',
        login: 'parkerziegler',
        bio: 'Software Engineer @Formidable Labs',
        company: 'Formidable Labs',
        location: 'Seattle, WA',
        followers: {
          totalCount: 41,
        },
        repositories: {
          totalCount: 25,
        },
        websiteUrl: 'https://parkerziegler.com',
        organizations: {
          edges: [],
        },
      },
    },
  };

  it('renders the loading screen if data is loading', () => {
    const wrapper = shallow(<AuthorScreen data={{ loading: true }} />);
    expect(wrapper.find(ActivityIndicator)).toHaveLength(1);
  });

  it('renders information on the GitHub user', () => {
    const wrapper = shallow(<AuthorScreen {...props} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('RepositoryOwner')).toHaveLength(1);
    expect(wrapper.find('AuthorOverview')).toHaveLength(1);
  });
});
