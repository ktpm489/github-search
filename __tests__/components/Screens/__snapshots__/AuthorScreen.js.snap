// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthorScreen /> renders information on the GitHub user 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AuthorScreen
    data={
      Object {
        "user": Object {
          "avatarUrl": "https://github.com/parkerziegler/avatar.png",
          "bio": "Software Engineer @Formidable Labs",
          "company": "Formidable Labs",
          "followers": Object {
            "totalCount": 41,
          },
          "location": "Seattle, WA",
          "login": "parkerziegler",
          "name": "Parker Ziegler",
          "organizations": Object {
            "edges": Array [],
          },
          "repositories": Object {
            "totalCount": 25,
          },
          "url": "https://github.com/parkerziegler",
          "websiteUrl": "https://parkerziegler.com",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <RepositoryOwner
          avatarUrl="https://github.com/parkerziegler/avatar.png"
          containerStyle={
            Object {
              "marginBottom": 10,
            }
          }
          flexDirection="column"
          height={100}
          infoContainerStyle={
            Object {
              "alignItems": "center",
              "marginBottom": 20,
            }
          }
          login="parkerziegler"
          name="Parker Ziegler"
          onAvatarPress={[Function]}
          url="https://github.com/parkerziegler"
          width={100}
        />,
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontSize": 18,
              "marginBottom": 20,
            }
          }
        >
          Software Engineer @Formidable Labs
        </Text>,
        <AuthorOverview
          company="Formidable Labs"
          followerCount={41}
          location="Seattle, WA"
          organizations={Array []}
          repoCount={25}
          url="https://github.com/parkerziegler"
          websiteUrl="https://parkerziegler.com"
        />,
      ],
      "style": Object {
        "backgroundColor": "#fff",
        "flex": 1,
        "justifyContent": "space-around",
        "padding": 20,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "avatarUrl": "https://github.com/parkerziegler/avatar.png",
          "containerStyle": Object {
            "marginBottom": 10,
          },
          "flexDirection": "column",
          "height": 100,
          "infoContainerStyle": Object {
            "alignItems": "center",
            "marginBottom": 20,
          },
          "login": "parkerziegler",
          "name": "Parker Ziegler",
          "onAvatarPress": [Function],
          "url": "https://github.com/parkerziegler",
          "width": 100,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "accessible": true,
          "allowFontScaling": true,
          "children": "Software Engineer @Formidable Labs",
          "ellipsizeMode": "tail",
          "style": Object {
            "color": "#000",
            "fontSize": 18,
            "marginBottom": 20,
          },
        },
        "ref": null,
        "rendered": "Software Engineer @Formidable Labs",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "company": "Formidable Labs",
          "followerCount": 41,
          "location": "Seattle, WA",
          "organizations": Array [],
          "repoCount": 25,
          "url": "https://github.com/parkerziegler",
          "websiteUrl": "https://parkerziegler.com",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <RepositoryOwner
            avatarUrl="https://github.com/parkerziegler/avatar.png"
            containerStyle={
              Object {
                "marginBottom": 10,
              }
            }
            flexDirection="column"
            height={100}
            infoContainerStyle={
              Object {
                "alignItems": "center",
                "marginBottom": 20,
              }
            }
            login="parkerziegler"
            name="Parker Ziegler"
            onAvatarPress={[Function]}
            url="https://github.com/parkerziegler"
            width={100}
          />,
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#000",
                "fontSize": 18,
                "marginBottom": 20,
              }
            }
          >
            Software Engineer @Formidable Labs
          </Text>,
          <AuthorOverview
            company="Formidable Labs"
            followerCount={41}
            location="Seattle, WA"
            organizations={Array []}
            repoCount={25}
            url="https://github.com/parkerziegler"
            websiteUrl="https://parkerziegler.com"
          />,
        ],
        "style": Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "space-around",
          "padding": 20,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "avatarUrl": "https://github.com/parkerziegler/avatar.png",
            "containerStyle": Object {
              "marginBottom": 10,
            },
            "flexDirection": "column",
            "height": 100,
            "infoContainerStyle": Object {
              "alignItems": "center",
              "marginBottom": 20,
            },
            "login": "parkerziegler",
            "name": "Parker Ziegler",
            "onAvatarPress": [Function],
            "url": "https://github.com/parkerziegler",
            "width": 100,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "accessible": true,
            "allowFontScaling": true,
            "children": "Software Engineer @Formidable Labs",
            "ellipsizeMode": "tail",
            "style": Object {
              "color": "#000",
              "fontSize": 18,
              "marginBottom": 20,
            },
          },
          "ref": null,
          "rendered": "Software Engineer @Formidable Labs",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "company": "Formidable Labs",
            "followerCount": 41,
            "location": "Seattle, WA",
            "organizations": Array [],
            "repoCount": 25,
            "url": "https://github.com/parkerziegler",
            "websiteUrl": "https://parkerziegler.com",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
